phpunit tests converted to mocha/chai

abaevbog opened the pull request:
Addresses [Issue #138](https://github.com/zotero/dataserver/issues/138)

v1 and v2 tests that do not require s3 integrations.

abaevbog committed:
phpunit test v1 and v2 converted to mocha/chai

abaevbog committed:
helper functions for api v3, annotation, atom and creator tests for v3

abaevbog committed:
v3 collection, group, fullText and note test

abaevbog committed:
v3 object test

abaevbog committed:
leftover object tests, mappings, params tests

abaevbog committed:
previously skipped tests requiring citation or translation server

abaevbog committed:
file tests v2 and half of v3

abaevbog committed:
the rest of file tests

abaevbog committed:
publication tests

abaevbog committed:
minor adjustments to prevent random 500 errors

dstillman commented on tests/remote_js/data/sync1download.xml:
sync1download.xml and sync2upload.xml can be removed. These were from the old sync system.

abaevbog commented on tests/remote_js/test/2/fileTest.js:
```diff
@@ -0,0 +1,703 @@
+		assert.equal(contentType, json.contentType);
+		assert.equal(charset, json.charset);
+		const updated = Helpers.xpathEval(xml, '/atom:entry/atom:updated');
+		assert.notEqual(serverDateModified, updated);
```
This line fails. In phpunit test, this condition "should" always fail as well.

There seems to be an error with xpath in phpunit on this [line](https://github.com/zotero/dataserver/blob/17e67f46061d28b6cf080ea48215b9477f6472e9/tests/remote/tests/API/2/FileTest.php#L292): `$serverDateModified = array_get_first($xml->xpath('/atom:entry/atom:updated'));`. It makes `serverDateModified` always null, so on this [line](https://github.com/zotero/dataserver/blob/17e67f46061d28b6cf080ea48215b9477f6472e9/tests/remote/tests/API/2/FileTest.php#LL383C1-L383C1), `$this->assertNotEquals($serverDateModified, array_get_first($xml->xpath('/atom:entry/atom:updated')));` always passes because one of the values is null.

Fixing the xpath in phpunit for `//atom:entry/atom:updated` makes `serverDateModified ` not null but then it is the same as `updated` value fetched from the xml, so the assertion fails.

So this line errors, just like it should in phpunit. I notice this condition was removed in V3. Should re remove it here as well?

abaevbog committed:
more v3 tests

abaevbog committed:
removing work dir, removed unneeded data/ files

abaevbog committed:
notification, tag and version tests

dstillman commented on tests/remote_js/api3.js:
```diff
@@ -0,0 +1,933 @@
+
+	static async resetSchemaVersion() {
+		const schema = JSON.parse(fs.readFileSync("../../htdocs/zotero-schema/schema.json"));
+		this.schemaVersion = schema;
```
This appears to be storing the entire file, not the schema version.

dstillman commented on tests/remote_js/package.json:
```diff
@@ -0,0 +1,19 @@
+{
+  "dependencies": {
+    "@aws-sdk/client-s3": "^3.338.0",
+    "axios": "^1.4.0",
```
Axios is gone, right?

dstillman reviewed (commented):
Still going through this, but some initial comments

abaevbog committed:
v3 itemTest

abaevbog committed:
some cleanup, proper retry logic when socket hangs

abaevbog committed:
using node-config

abaevbog committed:
removed inheritance from between api files and helpers

dstillman commented on tests/remote_js/test/shared.js:
```diff
@@ -0,0 +1,51 @@
+	},
+	API1WrapUp: async () => {
+		await API.userClear(config.userID);
+	},
```
Let's use `Before` and `After` like Mocha, `SetUp` and `TearDown` like PHPUnit, `Start`/`Stop`, `Start`/`Shutdown`, or something else consistent and distinct.

(`Setup` is a noun, so would need to be `SetUp` to be consistent, but `SetUp` and `WrapUp` are a bit too similar.)

abaevbog replied:
I renamed them for `API#Before` and `API#After`

abaevbog commented on tests/remote_js/api3.js:
```diff
@@ -0,0 +1,933 @@
+const fs = require("fs");
+const wgxpath = require('wgxpath');
+
+class API3 extends API2 {
```
Right, I did the extension to avoid duplicate code at first but since it is not an issue, I made independent api2/3 and helpers2/3 that have some functions overlapping

abaevbog commented on tests/remote_js/config.js:
```diff
@@ -0,0 +1,10 @@
+config = JSON.parse(data);
+config.timeout = 60000;
+config.numOwnedGroups = 3;
+config.numPublicGroups = 2;
```
Nothing specific here. I think the config.ini used in phpunit did not have those fields, and they were set in some of the other php files. I was converting files one by one, so it just turned out that way. 
Anyways, after my cleanup, I removed it all and setup the node-config with default.json5

abaevbog committed:
package.json

abaevbog commented on tests/remote_js/httpHandler.js:
```diff
@@ -0,0 +1,95 @@
+				if (error.name === 'FetchError') {
+					console.log('Request aborted. Wait for 2 seconds and retry...');
+					await new Promise(r => setTimeout(r, 2000));
+					tried += 1;
```
So I've been having the issue with the socket hanging up every now and then when I run test files. Happened randomly on different tests. 
This issue never happened with php-unit, so I wasn't sure what to think. I added this wrapper to be able to continue working on tests without getting stuck on figuring out the dataserver configuration.

Now that the tests are virtually done, I will dig into apache config on the zotero-dataserver. It must be something from there

abaevbog replied:
This actually was not a backend issue but a [bug in node-fetch](https://github.com/node-fetch/node-fetch/issues/1735).
I added a 0 second timeout before each call as suggested in the issue, and it seems to have solved it

abaevbog committed:
changed 'Setup' for 'Before' and 'Wrapup' for 'After' in shared.js, general test

abaevbog committed:
removing retries, using 0-second delay instead

dstillman commented on tests/remote_js/httpHandler.js:
```diff
@@ -27,26 +27,9 @@ class HTTP {
-		if (!success) {
-			throw new Error(`${method} to ${url} did not succeed after ${attempts} attempts.`);
-		}
+		await new Promise(resolve => setTimeout(resolve, 0));
```
You should add a comment with a link to the GitHub issue.

dstillman replied:
I just meant to comment the line so we know why this is here, not to add a comment on GitHub.

abaevbog replied:
Got it! My bad

dstillman replied:
I mean literally add a comment to the source code above this line and commit it. Nothing to do with our GitHub. There just shouldn’t be random unexplained lines in the source code that will stay there forever because we don’t know why they’re there.

abaevbog replied:
Understood. Sorry about that

abaevbog commented on tests/remote_js/httpHandler.js:
```diff
@@ -0,0 +1,78 @@
+			url = url.replace(localIPRegex, 'localhost');
+		}
+
+		await new Promise(resolve => setTimeout(resolve, 0));
```
`node-fetch` has a [bug](https://github.com/node-fetch/node-fetch/issues/1735) that causes socket to hang while running mocha tests.
Temporary solution is to add this `await new Promise(resolve => setTimeout(resolve, 0));` in `httpHandler` before executing each api call

abaevbog committed:
minor cleanup, added missing comments, some tweaking with socker hanging

abaevbog commented on tests/remote_js/httpHandler.js:
```diff
@@ -0,0 +1,95 @@
+// due to bug in node-fetch: https://github.com/node-fetch/node-fetch/issues/1735
+const httpAgent = new http.Agent({ keepAlive: true });
+const httpsAgent = new https.Agent({ keepAlive: true });
+const agentSelector = function (_parsedURL) {
```
Despite adding a tiny delay before each API call, the socket hanging happened again tonight as I was running tests. I added http agents with `keepalive=true `as another workaround suggested in the [issue](https://github.com/node-fetch/node-fetch/issues/1735). Seems to be helping but I will keep track on this. Thinking to try out axios just so see if it would help

abaevbog committed:
using helper functions for statusForObject + removed unnecessary parts from httpHandler

abaevbog committed:
initial pdf text extraction setup

abaevbog committed:
pdf text extraction test

abaevbog committed:
full-text tests with indexing function invoked locally

abaevbog committed:
items/unfiled and items/unfiled/tags tests

For #152

abaevbog committed:
?sort=editBy for group libraries

test for #154

abaevbog committed:
test to rename tag for PR #150

abaevbog committed:
test for literal || escaping for tags PR #143

abaevbog committed:
new annotation types test

abaevbog committed:
minor cleanup, skip tests for PRs, missing tests

Convert phpunit test updates to mocha
Skipping tests depending on not merged PRs
Fix to failing file test due to wrong content size

abaevbog committed:
skip annotation tags test for collections

abaevbog committed:
should_add_zero_integer_value_for_lastPageIndex

New converted test.
